src/main/java/org/opensearch/security/auditlog/sink/InternalOpenSearchDataStreamSink.java

```diff
@@ -0,0 +1,190 @@
+import org.opensearch.client.IndicesClient;
+import org.opensearch.client.indices.CreateDataStreamRequest;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.common.util.concurrent.ThreadContext.StoredContext;
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.security.auditlog.impl.AuditMessage;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.security.support.HeaderHelper;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.core.common.bytes.BytesArray;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.ResourceAlreadyExistsException;
+import org.opensearch.transport.RemoteTransportException;
+
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+public final class InternalOpenSearchDataStreamSink extends AuditLogSink {
+
+    private final Client clientProvider;
+    final String dataStreamName;
+    private final ThreadPool threadPool;
+    private boolean dataStreamInitialized = false;
+
+    public InternalOpenSearchDataStreamSink(
+        final String name,
+        final Settings settings,
+        final String settingsPrefix,
+        final Path configPath,
+        final Client clientProvider,
+        ThreadPool threadPool,
+        AuditLogSink fallbackSink
+    ) {
+        super(name, settings, settingsPrefix, fallbackSink);
+        this.clientProvider = clientProvider;
+        Settings sinkSettings = getSinkSettings(settingsPrefix);
+        this.threadPool = threadPool;
+
+        this.dataStreamName = sinkSettings.get(ConfigConstants.SECURITY_AUDIT_OPENSEARCH_DATASTREAM_NAME, "opensearch-security-auditlog");
+
+        // Node is no ready yet... this.initDataStream() must be called later (in method doStore())
+
+    }
+
+    private boolean initDataStream() {
+
+        if (this.dataStreamInitialized) {
+            return true;
+        }
+
+        Settings sinkSettings = getSinkSettings(settingsPrefix);
+
+        final boolean templateManage = sinkSettings.getAsBoolean(ConfigConstants.SECURITY_AUDIT_OPENSEARCH_DATASTREAM_TEMPLATE_MANAGE, true);
+
+        // Create datastream template
+        if (templateManage) {
+        
+            final String templateName = sinkSettings.get(ConfigConstants.SECURITY_AUDIT_OPENSEARCH_DATASTREAM_TEMPLATE_NAME, "opensearch-security-auditlog");
+            final Integer numberOfReplicas = sinkSettings.getAsInt(ConfigConstants.SECURITY_AUDIT_OPENSEARCH_DATASTREAM_TEMPLATE_NUMBER_OF_REPLICAS, 0);
+            final Integer numberOfShards = sinkSettings.getAsInt(ConfigConstants.SECURITY_AUDIT_OPENSEARCH_DATASTREAM_TEMPLATE_NUMBER_OF_SHARDS, 1);
+
+            final String auditlogTemplateJson = "{"
+                + "\"index_patterns\": [ \""+ dataStreamName +"\" ],"
```

Comment:
Is there a Builder pattern somewhere that can help assemble this payload?

I haven't looked at it in detail, but something like this: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/action/admin/indices/datastream/DataStreamTestCase.java#L81-L89

Reply:
Yes, that's really ugly, but i found no other way. Thanks for the link, i fixed it.